fortran/testpar/async.F90

```diff
@@ -0,0 +1,1382 @@
+
+  END SUBROUTINE H5O_async_tests
+
+END MODULE test_async_APIs
+
+!
+! The main program for async HDF5 Fortran tests
+!
+PROGRAM async_test
+  USE, INTRINSIC :: ISO_C_BINDING, ONLY : C_INT64_T
+  USE HDF5
+  USE MPI
+  USE TH5_MISC
+  USE TH5_MISC_GEN
+  USE test_async_APIs
+
+  IMPLICIT NONE
+
+  INTEGER :: total_error = 0        ! sum of the number of errors
+  INTEGER :: mpierror               ! MPI hdferror flag
+  INTEGER :: mpi_size               ! number of processes in the group of communicator
+  INTEGER :: mpi_rank               ! rank of the calling process in the communicator
+  INTEGER :: required, provided
+
+  INTEGER(HID_T) :: vol_id
+  INTEGER :: hdferror
+  LOGICAL :: registered
+  INTEGER :: sum
+  INTEGER :: nerrors = 0
+
+  LOGICAL          :: cleanup
+  INTEGER          :: ret_total_error = 0
+
+  !
+  ! initialize MPI
+  !
+  required = MPI_THREAD_MULTIPLE
+  CALL mpi_init_thread(required, provided, mpierror)
+  IF (mpierror .NE. MPI_SUCCESS) THEN
+     WRITE(*,*) "MPI_INIT_THREAD  *FAILED*"
+     nerrors = nerrors + 1
+  ENDIF
+  IF (provided .NE. required) THEN
+     total_error = -1 ! skip test
+     IF(mpi_rank==0) CALL write_test_status(total_error, &
+          "MPI doesn't support MPI_Init_thread with MPI_THREAD_MULTIPLE", total_error)
+     STOP
+  ENDIF
+  CALL mpi_comm_rank( MPI_COMM_WORLD, mpi_rank, mpierror )
+  IF (mpierror .NE. MPI_SUCCESS) THEN
+     WRITE(*,*) "MPI_COMM_RANK  *FAILED* Process = ", mpi_rank
+     nerrors = nerrors + 1
+  ENDIF
+  CALL mpi_comm_size( MPI_COMM_WORLD, mpi_size, mpierror )
+  IF (mpierror .NE. MPI_SUCCESS) THEN
+     WRITE(*,*) "MPI_COMM_SIZE  *FAILED* Process = ", mpi_rank
+     nerrors = nerrors + 1
+  ENDIF
+
+  IF(nerrors.NE.0)THEN
+     IF(mpi_rank==0) CALL write_test_status(sum, &
+          'Testing Initializing mpi_init_thread', total_error)
+     CALL MPI_Barrier(MPI_COMM_WORLD, mpierror)
+     CALL mpi_abort(MPI_COMM_WORLD, 1, mpierror)
+  ENDIF
+
+  !
+  ! Initialize the HDF5 fortran interface
+  !
+  CALL h5open_f(hdferror)
+
+  ! CHECK ASYNC VOLS AVAILABILITY
+  !
+  ! (1) Check if ASYNC VOL is available
+  CALL H5VLis_connector_registered_by_name_f("async", registered,  hdferror)
+  CALL check("H5VLis_connector_registered_by_name_f", hdferror, total_error)
+
+  IF(.NOT.registered)THEN
+
+     ! (2) check if the DAOS VOL is available
+     CALL H5VLis_connector_registered_by_name_f("daos", registered,  hdferror)
+     CALL check("H5VLis_connector_registered_by_name_f", hdferror, total_error)
+
+     IF(.NOT.registered)THEN
+        ! No async compatible VOL found, skipping test
+        total_error = -1
+     ELSE
+        CALL H5Vlregister_connector_by_name_f("daos", vol_id, hdferror)
+        CALL check("H5Vlregister_connector_by_name_f", hdferror, total_error)
+     ENDIF
+
+  ELSE
+     CALL H5Vlregister_connector_by_name_f("async", vol_id, hdferror)
+     CALL check("H5Vlregister_connector_by_name_f", hdferror, total_error)
+  ENDIF
```

Comment:
Not that they would be testing anything in particular, but you should be able to always run the async tests since they should just act as passthrough tests for VOLs that don't support async.

Reply:
Yes, but all the ES count checks would be wrong.